scipy/_lib/doccer.py

```diff
@@ -1,26 +1,39 @@
-''' Utilities to allow inserting docstring fragments for common
-parameters into function and method docstrings'''
+"""Utilities to allow inserting docstring fragments for common
+parameters into function and method docstrings."""
 
+from collections.abc import Callable, Iterable
+from typing import Any, ParamSpec, TypeAlias, TypeVar
 import sys
 
 __all__ = [
-    'docformat', 'inherit_docstring_from', 'indentcount_lines',
-    'filldoc', 'unindent_dict', 'unindent_string', 'extend_notes_in_docstring',
-    'replace_notes_in_docstring', 'doc_replace'
+    "docformat",
+    "inherit_docstring_from",
+    "indentcount_lines",
+    "filldoc",
+    "unindent_dict",
+    "unindent_string",
+    "extend_notes_in_docstring",
+    "replace_notes_in_docstring",
+    "doc_replace",
 ]
 
+P = ParamSpec("P")
+R = TypeVar("R")
 
-def docformat(docstring, docdict=None):
-    ''' Fill a function docstring from variables in dictionary
+Decorator: TypeAlias = Callable[[Callable[..., R]], Callable[..., R]]
```

Comment:
The parameter signatures of the passed functions will get lost here. 
If this decorator type doesn't change the function signature (by e.g. changing the type of some parameter), then you instead could use something like 

```suggestion
_F = TypeVar("_F", bound=Callable[..., object])

Decorator: TypeAlias = Callable[[_F], _F]
```

The reason for binding the entire callable, and not using `ParamSpec`, is that this way the entire function type is bound, including all overloads, and possible attributes. So this way, e.g. decorating a `np.ufunc` would return that same `np.ufunc` type, instead of effectively upcasting it to a `Callable[..., object]`.

Reply:
Unless I'm missing something, if you apply this "trick" to other similar cases in this module, this would allow you to get rid of the `ParamSpec`